src/irmin-pack/pack_store.ml

```diff
@@ -222,36 +396,50 @@ module Maker
 
     let auto_flush = 1024
 
-    let unsafe_append ~ensure_unique ~overcommit t k v =
-      if ensure_unique && unsafe_mem t k then ()
-      else (
-        [%log.debug "[pack] append %a" pp_hash k];
-        let offset k =
-          match Index.find t.pack.index k with
-          | None ->
-              Stats.incr_appended_hashes ();
-              None
-          | Some (off, _, _) ->
-              Stats.incr_appended_offsets ();
-              Some off
+    let unsafe_append ~ensure_unique_indexed ~overcommit t hash v =
+      let unguarded_append () =
+        [%log.debug "[pack] append %a" pp_hash hash];
+        let offset_of_key k =
+          (* XXX: don't need option any more? *)
+          Stats.incr_appended_offsets ();
+          Some (Key.to_offset k)
```

Comment:
The key may be indirect, that call is unsafe.

For example, a freshly snapshot-imported irmin-pack contains a few child pointers by hash

Reply:
Indeed, thanks. This is a hold-over from when all keys were direct.

I've changed the implementation to query the index when the key is indirect, and removed the `Key.to_{offset,length}` functions altogether to make sure it doesn't happen again.